### PR TITLE
Reflection free code gen

### PR DESCRIPTION
**Pull request by dsyme**
There is a strong desire in the .NET and F# community to set up .NET and F# to be more friendly to native compilation tool chains and "tree shaking" compilation. 

See also https://github.com/dotnet/fsharp/issues/12819 and https://github.com/dotnet/fsharp/issues/11891

Historically this has been a tricky area for both F# and .NET generally, some of the history is mentioned here: https://github.com/fsharp/fslang-suggestions/issues/919.  Looking forward, there are recent strong efforts in this area led by @jkotas and others.

Now, this is as an area where "small but deadly" things can kill you - one generated construct that a native toolchain can't handle. or one reference through to some large chunk of library code that gets linked in. Problems we definitely know of are

1. F# currently implicitly emits implementations of `ToString()` for records, unions and structs (also `get_Message()` for exception definitions) and these in turn use `sprintf` with `%A` formatting.   F# also emits DebuggerDisplayAttribute for union types that do likewise.

2. Users may use `%A` formatting 

3. Users may use quotations (either implicitly or explicitly)

4. The user may end up using libraries that use these constructs.

The proposal in https://github.com/fsharp/fslang-suggestions/issues/919 is to address these by add a command line option `--reflectionfree` that adjusts F# code generation to both avoid the generation of the problematic constructs, and to give an error if `%A` is used.

Note that (1) is not a problem in FSharp.Core as the code generation is already explicitly suppressed in that case

(2) is potentially a problem - this PR removes a few explicit uses of `%A` in FSharp.Core  

(3) is probably better dealt with by the user simply avoiding quotations, but not banning their use outright.

Note that using this option would affect both the ToString semantics (no implicit implementation is given) and the debugging experience 

* [ ] try it
* [ ] check the uses of reflection in FSharp.Core where we look at attributes of types to determine equality semantics
* [ ] check uses of reflection in FSharp.Core related to dynamic implementations of some operators.

### CODE REVIEW

File: src/fsharp/CheckFormatStrings.fs [hunk 1]
```diff
@@ -435,6 +444,9 @@ let parseFormatStringInternal (m: range) (fragRanges: range list) (g: TcGlobals)
                   parseLoop ((posi, NewInferenceType g) :: acc) (i, fragLine, startFragCol) fragments
 
               | 'A' ->
+                  if g.useReflectionFreeCodeGen then
+                      failwithf "%s" (FSComp.SR.forPercentAInReflectionFreeCode())
```
Comment (inosik) [inline src/fsharp/CheckFormatStrings.fs:448]: Did you mean this?

```suggestion
                      failwith (FSComp.SR.forPercentAInReflectionFreeCode())
```

**Comment by kerams** [comment]
> Note that using this option would affect both the ToString semantics (no implicit implementation is given)

Could this effort be paired with emitting specialized `ToString` methods and https://github.com/fsharp/fslang-suggestions/issues/1108? Or one step at a time? :)

**Comment by dsyme** [comment]
@jkotas writes

> We have a section on trimming and NativeAOT in our regular preview announcement blog post. I think we should mention this is a good example of how .NET ecosystem is approaching trimming and AOT compatibility.


>
> We depend heavily on the IL static analysis and annotations to detect reflection patterns that may break with AOT or linker. It is documented in [https://docs.microsoft.com/en-us/dotnet/core/deploying/trimming/prepare-libraries-for-trimming](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fdocs.microsoft.com%2Fen-us%2Fdotnet%2Fcore%2Fdeploying%2Ftrimming%2Fprepare-libraries-for-trimming&data=05%7C01%7CDon.Syme%40microsoft.com%7C537a2bc55a7449b7e81f08da191d4bfb%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637849910952308282%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=%2Bio5%2BsMbMJ0ruyu7MoG2L4UficZGrbatEIkALnCyGl4%3D&reserved=0) . We have the static analysis implemented both as Roslyn analyzer (not applicable for F#) and as part of trimmer/AOT compiler. You should be able to verify that you got all problematic reflection patterns covered by getting zero static analysis warnings when compiling sample F# apps.

**Comment by vzarytovskii** [comment]
> > Note that using this option would affect both the ToString semantics (no implicit implementation is given)
> 
> Could this effort be paired with emitting specialized `ToString` methods and [fsharp/fslang-suggestions#1108](https://github.com/fsharp/fslang-suggestions/issues/1108)? Or one step at a time? :)

I'd personally keep it separate features/contributions.

**Comment by agocke** [comment]
This is really cool! If you want to verify that your app is safe for AOT and trimming, the linker can produce trim warnings whenever there is code that is not statically understood. Instructions are available at https://docs.microsoft.com/en-us/dotnet/core/deploying/trimming/trim-self-contained

Looking forward, I think the biggest trimming concern that I would have in F# is serialization. Unfortunately, most serialization patterns are simply too complex to statically annotate. The recommended substitution is to use a source-generated serializer, like the JSON source generators or my [serde-dn](https://github.com/agocke/serde-dn) library. Unfortunately, those are reliant on C# source generators in their API construction (serde-dn works by having the source generator implement the serialization interfaces automatically), so some way of providing similar functionality in F# would be very interesting.

**Comment by baronfel** [comment]
The clearest parallel for F# to source generators IMO are generative Type Providers, which are compiler-hosted components that emit code  mid-compilation. There are some rough edges around the UX for them, and debugging the generated code as you are writing one of them can be a little gnarly, but the overall mechanism is pretty well understood.

**Comment by agocke** [comment]
Yup, I'll just speak for myself here since I don't know exactly what the flexibility around JSON source generation is: serde-dn operates fundamentally around a series of interfaces that ensure type safety of serialization, ease of customization, and flexibility of format. In general, the preferred pattern for types that appear in source is to implement the appropriate interfaces on the types directly. There are always cases where you can't implement an interface directly (like on a type you don't own) and there is a system for accommodating that (generating wrapper structs), but the model really prefers implementing interfaces directly when possible.

My understanding is that type providers wouldn't provide that capability, but I might be out of date here.

**Comment by dsyme** [comment]
> Looking forward, I think the biggest trimming concern that I would have in F# is serialization. 

Yes.  Realistically I'd expect it to play out like this:

* Short term:
  1. Most F# users who care about trimming down (relatively rare) use and develop F#-specific code generation for serializers, e.g. Myriad.  Alternatively they simply hand-write using serializer combinators, which is pretty pleasant in F# and is a good way of "owning your own destiny"
  2. A small portion work to make sure type provider solutions like FSharp.Data are capable of doing this (for F# types defined in another assembly)
  3. A small portion use C# source generators in C# projects over a C# object model. 

* Medium term:
  1. The above techniques mature
  2. F# type providers get adapted [to allow them to work over types defined in the current assembly](https://github.com/fsharp/fslang-suggestions/issues/212), which is approved in principle but a substantial ticket.  We'd then expect F#-specific libraries like FSharp.Data to handle serialization.

**Comment by dsyme** [comment]
> ...generative type providers...

TBH I think erased type provides are also pretty feasible - see notes above on FSharp.Data today.

**Comment by agocke** [comment]
@dsyme Very cool. Yup, FSharp.Data looks like it should work just fine with trimming.

**Comment by dsyme** [comment]
Linking this: https://github.com/fsharp/fslang-suggestions/issues/919#issuecomment-1136951429

**Comment by dsyme** [comment]
@0101 asked what next steps should be here

I think the first thing to determine is "does it matter if some things in FSharp.Core use reflection, and what should we do about it"?

Specific things

* Is `ty.GetCustomAttributes(typeof<CompilationMappingAttribute>, false)` etc.  used in FSharp.Core considered "reflection-free" for the purposes of trimming, .NET native etc.  I'm hopeful that this is not the case and that basic reflection on the custom attributes of .NET types is permitted - that this doesn't count as "unreferenced code".  If not we have a significant problem because the semantics of F# generic equality rely on being able to do this kind of reflection (that is, we need to know a few bits of information about each type definition we encounter).  It's not totally the end-of-the-world if it is a problem but it would mean that trimmed code could change semantics in subtle ways.

* Should we be sprinkling [`RequiresUnreferencedCodeAttribute`](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.requiresunreferencedcodeattribute?view=net-6.0) over the things that use reflection constructs in a way that kills trimming, including

  - Printf.printfn  and friends

  - ExtraTopLevelOperators.printfn and friends

  - FSharp.Reflection.* 

  - FSharp.Quotations.* 

  - FSharp.Linq.* 

  - ExtraTopLevelOperators.query

* Do the FSharp.Core entry points leading to UnaryDynamicImpl and BinaryDynamicImpl get trimmed out if they're never actually needed?  These are used for the "dynamic" implementations of `FSharp.Core.Operators.(+)` but because of inlining we don't actually emit calls to these in F# code except when generating the further dynamic implementations of generic inlined code that calls these, e.g. `let inline f x = x + x` generates a call to `FSharp.Core.Operators.(+)`

* Should we be putting `RequiresUnreferencedCodeAttribute` on `FSharp.Core.Operators.(+)` and friends for use by analysis tooling - again, the "usual" use of these operators in F# gets inlined and flattened, and the attributes would only be there for the reflective code emitted in codegen.

* Is Roslyn in any way aware of  `RequiresUnreferencedCodeAttribute` 

Overall it would be great to get community and expert advice on this, and also get an understanding for what will go wrong if it doesn't work. I did see some related issues where F# folk are starting to mark stuff as trimmable:

* https://github.com/fsprojects/FSharp.Data.Adaptive/issues/104

* https://github.com/fsprojects/FSharp.Control.Reactive/pull/166/files#diff-acb34de7cd05d0ab12f08b27c304e074405a6b0e54154697a9af071ec170d1d9R5

* https://github.com/fsprojects/FSharp.Data.Adaptive/pull/105

* https://github.com/fsprojects/FSharp.Data.Adaptive/pull/105/files#diff-79c74b83b5f8725083b39f21f41b468e68ba946d669475b17472e5ae97fa09d7R89

**Comment by vzarytovskii** [comment]
> * Should we be sprinkling [`RequiresUnreferencedCodeAttribute`](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.requiresunreferencedcodeattribute?view=net-6.0) over the things that use reflection constructs in a way that kills trimming, including

I guess there are several parts of this - reflection-free codegen by compiler itself ("%A", quotations, dynamic imlpementations, etc), reducing/reviewing reflection in FSharp.Core and making FSharp.Core adjustments for easier trimmability.

**Comment by jkotas** [comment]
> Is ty.GetCustomAttributes(typeof<CompilationMappingAttribute>, false) etc. used in FSharp.Core considered "reflection-free" for the purposes of trimming,

Reading custom attributes is compatible with trimming and AOT compilation.

(I would not call it "reflection-free". Custom attributes are reflection by design.)

> Should we be sprinkling RequiresUnreferencedCodeAttribute over the things that use reflection constructs in a way that kills trimming,

RequiresUnreferencedCodeAttribute allows static analyzers to provide better diagnostic messages. 

For example, if you do not sprikle RequiresUnreferencedCodeAttribute over `Printf.printfn` and somebody tries to trim the application that calls `Printf.printfn`, they will see warnings about trim unfriendly APIs being called somewhere inside FSharp.Core implementation, it won't be straightforward for them to figure out that these warnings are caused by a call to `Printf.printfn`.

If you sprinkle RequiresUnreferencedCodeAttribute over `Printf.printfn` and somebody tries to trim the application that calls `Printf.printfn`, they will see warning about the `Printf.printfn`, making the problem much more obvious.

> Is Roslyn in any way aware of RequiresUnreferencedCodeAttribute

Roslyn itself is not aware of RequiresUnreferencedCodeAttribute. 

> What analyzer support does C# by default for all this stuff. 

We have Roslyn analyzer that checks RequiresUnreferencedCodeAttribute and friends in the code that you are compiling from source. This analyzer is good for dev inner loop. You would have to reimplement this analyzer for F#. 

The IL linker and AOT compiler have same analyzer that works on the IL of the whole app. You will get this analyzer for free for F#.

**Comment by dsyme** [comment]
@jkotas Thank you!!

**Review by 0101 — APPROVED** [review]

**Comment by vzarytovskii** [comment]
I'd say we should merge this first set of changes and then continue working on corelib from there.

**Review by vzarytovskii — APPROVED** [review]

**Pull request merged**